### ISSUE TITLE
Should the error mock actually throw an exception?

### ISSUE THREAD

**Issue by nre-ableton** [jenkinsci/JenkinsPipelineUnit#176]
So in https://github.com/jenkinsci/JenkinsPipelineUnit/pull/175, I originally had a commit which changed the `error` mock to this:

```groovy
helper.registerAllowedMethod("error", [String.class]) { String message ->
    updateBuildStatus('FAILURE')
    throw new Exception(message)
}
```

IMHO, this should be the desired behavior, otherwise a pipeline like this will not execute correctly with this framework:

```groovy
node('whatever') {
    stage('example') {
        try {
            if (!fileExists('foo')) {
                error 'something went terribly wrong'
            }
        } catch (error) {
            sh 'report-error-to-server.sh'
            throw error
        }

        sh 'business-as-usual.sh'
    }
}
```

So in the above example, if the user calls `assertJobStatusFailure()` they will see that the pipeline failed, but the work below the `error` will still be done. This is not how Jenkins would behave, as it would throw `hudson.AbortException`. So if the user wants to assert that `sh 'business-as-usual.sh'` _was not called_ in this case, that test would fail. Likewise, if the user wants to assert that the `sh 'report-error-to-server.sh'` step was called, that would not work.

For these reasons, I believe that `error` should throw an exception in its mock implementation. However, several tests rely on the previous behavior of _not_ throwing (for example: https://github.com/jenkinsci/JenkinsPipelineUnit/blob/master/src/test/groovy/com/lesfurets/jenkins/TestSharedLibraryWithProjectSourceRetriever.groovy#L81). For this reason, I am disinclined to propose such a change in a PR without understanding why the current behavior is the way that it is. Can someone shed some light on the situation?

**Comment by stchar** [comment]
I'm not sure, I would prefer default mock just to update the the pipeline status.
When I need to verify exception I do it with exactly `thrown` object for a particular testcase, like it described in the [README](https://github.com/jenkinsci/JenkinsPipelineUnit/blob/master/README.md#check-pipeline-exceptions)

There are some caveats:
1. Backward compatibility: I did it once, it was a big-bang, many testcases were not ready, just because they were designed to verify slightly different areas of logic. `error` was triggered by a shared library which had to be mocked properly for each case. So I turned it off.
2. Parallel interceptor needs to be wrapped in try catch if it was called with `failFast:false`
